internal/datanode/data_node.go

```diff
@@ -820,6 +819,39 @@ func (node *DataNode) Compaction(ctx context.Context, req *datapb.CompactionPlan
 	}, nil
 }
 
+// GetCompactionState called by DataCoord
+// return status of all compaction plans
+func (node *DataNode) GetCompactionState(ctx context.Context, req *datapb.CompactionStateRequest) (*datapb.CompactionStateResponse, error) {
+	log.Info("DataNode.GetCompactionState")
+	results := make([]*datapb.CompactionStateResult, 0)
+	node.compactionExecutor.executing.Range(func(k, v any) bool {
+		results = append(results, &datapb.CompactionStateResult{
+			Status: &commonpb.Status{
+				ErrorCode: commonpb.ErrorCode_Success,
+			},
+			State:  commonpb.CompactionState_Executing,
+			PlanID: k.(UniqueID),
+		})
+		return true
+	})
+	node.compactionExecutor.completed.Range(func(k, v any) bool {
+		results = append(results, &datapb.CompactionStateResult{
+			Status: &commonpb.Status{
+				ErrorCode: commonpb.ErrorCode_Success,
+			},
+			State:  commonpb.CompactionState_Completed,
+			PlanID: k.(UniqueID),
+			Result: v.(*datapb.CompactionResult),
+		})
+		return true
+	})
+	node.compactionExecutor.cleanCompleted()
```

Comment:
This block may has concurrency problem. If a compaction plan completed between `iteration` and `cleanCompleted`, its result will be missed.